### PR TITLE
[IMP] functions: Add #SPILL! error

### PR DESCRIPTION
**Pull request by rrahir**
## Description:

This revision adds the #SPILL! error to give an insight to the user
about the type of error in the cell.
The cell itself or its formula are not broken or incorrect, they just
happen to be blocked, either by a lack of dimension in the sheet or
another cell content that prevent the formula to spread itself.

Task: : [3874364](https://www.odoo.com/web#id=3874364&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/types/errors.ts [hunk 1]
```diff
@@ -46,3 +47,9 @@ export class UnknownFunctionError extends EvaluationError {
     super(message, CellErrorType.UnknownFunction);
   }
 }
+
+export class SplillBLockedError extends EvaluationError {
```
Comment (LucasLefevre) [inline src/types/errors.ts:51]:
```suggestion
export class SplillBlockedError extends EvaluationError {
```

**Review by LucasLefevre — COMMENTED** [review]
I like the dashed highlight to show the spread zone, good idea :)

File: src/stores/array_formula_highlight.ts [hunk 2]
```diff
@@ -12,29 +13,25 @@ export class ArrayFormulaHighlight extends SpreadsheetStore {
...
       return [];
     }
-
-    const sheetId = this.model.getters.getActiveSheetId();
     return [
       {
-        sheetId,
+        sheetId: position.sheetId,
         zone,
+        dashed: !!(cell.value === CellErrorType.SpilledBlocked),
```
Comment (LucasLefevre) [inline src/stores/array_formula_highlight.ts:30]:
```suggestion
        dashed: cell.value === CellErrorType.SpilledBlocked,
```

File: src/stores/array_formula_highlight.ts [hunk 3]
```diff
@@ -12,29 +13,25 @@ export class ArrayFormulaHighlight extends SpreadsheetStore {
   }
 
   get highlights(): Highlight[] {
-    const zone = this.getHighlightZone();
+    let zone: Zone | undefined;
+    const position = this.model.getters.getActivePosition();
+    const cell = this.getters.getEvaluatedCell(position);
+    const spreader = this.model.getters.getArrayFormulaSpreadingOn(position);
+    zone = spreader
+      ? this.model.getters.getSpreadZone(spreader, true)
```
Comment (LucasLefevre) [inline src/stores/array_formula_highlight.ts:21]: I would force to name the boolean argument
```suggestion
      ? this.model.getters.getSpreadZone(spreader, { ignoreSpillError: true })
```

**Review by LucasLefevre — COMMENTED** [review]
Looking good
only 2 nitpicking suggestions

**Comment by rrahir** [comment]
@LucasLefevre updated

**Review by LucasLefevre — APPROVED** [review]
robodoo r+ rebase-ff

**Pull request closed without merging**